kikuchipy/indexation/similarity_metrics.py

```diff
@@ -0,0 +1,394 @@
+        self._dtype_out = dtype_out
+        self.sign = sign
+        self.flat = flat
+        self.scope = scope
+
+    def __call__(self, patterns, templates):
+        dtype = self._dtype_out
+        patterns = patterns.astype(dtype)
+        templates = templates.astype(dtype)
+        if isinstance(patterns, da.Array):
+            patterns = patterns.rechunk()
+        if isinstance(templates, da.Array):
+            templates = templates.rechunk()
+        if self._make_compatible_to_lower_scopes:
+            patterns, templates = self._expand_dims_to_match_scope(
+                patterns, templates
+            )
+        return self._measure(patterns, templates).squeeze()
+
+    def _measure(self, patterns, templates):
+        return self._metric_func(patterns, templates)
+
+    def _expand_dims_to_match_scope(self, p, t):
+        p_scope_ndim, t_scope_ndim = self._SCOPE_TO_P_T_NDMI[self.scope]
+        p = p[(np.newaxis,) * (p_scope_ndim - p.ndim)]
+        t = t[(np.newaxis,) * (t_scope_ndim - t.ndim)]
+        return p, t
+
+    def _is_compatible(self, p, t):
+        p_ndim, t_ndim = p.ndim, t.ndim
+        if self.flat:
+            p_ndim -= 2
+            t_ndim -= 1
+        inferred_scope = self._P_T_NDIM_TO_SCOPE.get((p_ndim, t_ndim), False)
+        if not inferred_scope:
+            return False
+        if inferred_scope == self.scope:
+            return True
+        else:
+            if self._make_compatible_to_lower_scopes:
+                return inferred_scope in self._SCOPE_TO_LOWER_SCOPES[self.scope]
+            else:
+                return False
+
+
+class FlatSimilarityMetric(SimilarityMetric):
+    """Similarity metric between 2D gray-tone images where the images are flattened before sent to `metric_func`"""
+
+    _P_T_NDIM_TO_SCOPE = {
+        (2, 2): MetricScope.MANY_TO_MANY,
+        (1, 2): MetricScope.ONE_TO_MANY,
+        (3, 1): MetricScope.MANY_TO_ONE,
+        (1, 1): MetricScope.ONE_TO_ONE,
+    }
+
+    _SCOPE_TO_P_T_NDMI = {
+        MetricScope.MANY_TO_MANY: (2, 2),
+        MetricScope.ONE_TO_MANY: (1, 2),
+        MetricScope.MANY_TO_ONE: (3, 1),
+        MetricScope.ONE_TO_ONE: (1, 1),
+    }
+
+    _PARTIAL_SCOPE_TO_NDIM = {
+        "MANY_": 2,  #  Many patterns
+        "_MANY": 2,  # Many templates
+        "ONE_": 1,  # One pattern
+        "_ONE": 1,  # One template
+    }
+
+    def _measure(
+        self,
+        patterns: Union[np.ndarray, da.Array],
+        templates: Union[np.ndarray, da.Array],
+    ):
+        nav_shape = _get_nav_shape(patterns)
+        sig_shape = _get_sig_shape(patterns)
+        nav_flat_size, sig_flat_size = np.prod(nav_shape), np.prod(sig_shape)
+        patterns = patterns.reshape((nav_flat_size, sig_flat_size))
+        templates = templates.reshape((-1, sig_flat_size))
+        return self._metric_func(patterns, templates)
+
+
+def expand_dims_to_many_to_many(p, t, flat):
```

Comment:
I suggest making this private, unless it's ment to be used directly by the user.